ddl/placement/bundle.go

```diff
@@ -49,6 +49,224 @@ func NewBundle(id int64) *Bundle {
+	if err != nil {
+		return nil, fmt.Errorf("%w: 'Constraints' should be [constraint1, ...] or any yaml compatible array representation", err)
+	}
+
+	LeaderConstraints, err := NewConstraintsFromYaml([]byte(leaderConstraints))
+	if err != nil {
+		return nil, fmt.Errorf("%w: 'LeaderConstraints' should be [constraint1, ...] or any yaml compatible array representation", err)
+	}
+	for _, cnst := range CommonConstraints {
+		if err := LeaderConstraints.Add(cnst); err != nil {
+			return nil, fmt.Errorf("%w: LeaderConstraints conflicts with Constraints", err)
+		}
+	}
+
+	Rules := []*Rule{}
+	if len(LeaderConstraints) > 0 {
+		Rules = append(Rules, NewRule(Leader, 1, LeaderConstraints))
+	}
+
+	if voterCount > 0 {
+		VoterRules, err := NewRules(Voter, voterCount, voterConstraints)
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid VoterConstraints", err)
+		}
+		for _, rule := range VoterRules {
+			for _, cnst := range CommonConstraints {
+				if err := rule.Constraints.Add(cnst); err != nil {
+					return nil, fmt.Errorf("%w: VoterConstraints conflicts with Constraints", err)
+				}
+			}
+		}
+		Rules = append(Rules, VoterRules...)
+	}
+
+	if followerCount > 0 {
+		FollowerRules, err := NewRules(Follower, followerCount, followerConstraints)
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid FollowerConstraints", err)
+		}
+		for _, rule := range FollowerRules {
+			for _, cnst := range CommonConstraints {
+				if err := rule.Constraints.Add(cnst); err != nil {
+					return nil, fmt.Errorf("%w: FollowerConstraints conflicts with Constraints", err)
+				}
+			}
+		}
+		Rules = append(Rules, FollowerRules...)
+	}
+
+	if learnerCount > 0 {
+		LearnerRules, err := NewRules(Learner, learnerCount, learnerConstraints)
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid LearnerConstraints", err)
+		}
+		for _, rule := range LearnerRules {
+			for _, cnst := range CommonConstraints {
+				if err := rule.Constraints.Add(cnst); err != nil {
+					return nil, fmt.Errorf("%w: LearnerConstraints conflicts with Constraints", err)
+				}
+			}
+		}
+		Rules = append(Rules, LearnerRules...)
+	}
+
+	return &Bundle{Rules: Rules}, nil
+}
+
+// NewBundleFromSugarOptions will transform syntax sugar options into the bundle.
+func NewBundleFromSugarOptions(primaryRegion string, regions []string, followers uint64, schedule string) (*Bundle, error) {
+	if followers == 0 {
+		followers = 2
+	}
+	var constraints Constraints
+	var err error
+
+	if len(primaryRegion) == 0 {
+		return nil, fmt.Errorf("%w: empty primary region", ErrInvalidPlacemeOptions)
+	}
+
+	Rules := []*Rule{}
+	switch strings.ToLower(schedule) {
+	case "", "even":
+		constraints, err = NewConstraints([]string{fmt.Sprintf("+region=%s", primaryRegion)})
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid PrimaryRegion '%s'", err, primaryRegion)
+		}
+		Rules = append(Rules, NewRule(Leader, 1, constraints))
+	case "majority_in_primary":
```

Comment:
@morgo PTAL, I want to ensure the behavior of "majority" is correct, or behave as your expectation.

Reply:
even num:  1(leader) + 2(follower) = should get the 2 out of 3 to quick vote 
odd num: 1(leader) + 3(follower)= should get the 3 out of 4 to quick vote

Reply:
I think we should use an integer value here so it is easier to follow:
```
    // number of quorum members, without the leader
    primary := (followers + 1) / 2
```

I.e. we need the ceil of followers/2 because we always have the leader as well:
```
// We already have the leader, so we need to calculate how many additional followers
// need to be in the primary region for quorum
followersInPrimary := int(math.Ceil( float64(followers) / 2))
```

Some example math:
2 followers: primary =1
3 followers: primary = 2 (* not recommended config)
4 followers: primary = 2

Reply:
Addressed.